### PR TITLE
Fix VM pool updates with revision tracking

### PR DESCRIPTION
**Pull request by davidvossel**
This PR fixes an issue that caused inconsistent VirtualMachinePool updates for VMs.

## The Bug
Previously, the vmpool controller calculated a hash of the pool's virtualmachine template and compared that hash to a value stored on the underlying VM/VMI. If the hash values changed, that meant a change occurred on the Pool's spec that needs to be applied to the underlying VM/VMIs.

The issue here is that the hash produced results that could vary release to release. The same pool spec could produce two different hashs depending on the kubevirt release currently in use. The reason for this is fields added to our api influence the hashed result regardless if the vm pool uses those fields or not.

This means that each kubevirt update runs the risk of causing a mass update to VM/VMIs if a new field is added to the VMI api.

## The Fix

Now instead of comparing hashes, we compare controller revisions. A controller revision of a pool is created and referenced by VM/VMIs. We perform a semantic DeepEqual's of the controller revision referenced by the VM/VMI with the latest controller revision for the vmpool in order to determine if an update needs to occur.

```release-note
Fixes inconsistent VirtualMachinePool VM/VMI updates by using controller revisions
```

### CODE REVIEW

File: pkg/virt-controller/watch/pool.go [hunk 1]
```diff
@@ -408,7 +441,7 @@ func (c *PoolController) Run(threadiness int, stopCh <-chan struct{}) {
 	log.Log.Info("Starting pool controller.")
 
 	// Wait for cache sync before we start the pool controller
-	cache.WaitForCacheSync(stopCh, c.poolInformer.HasSynced, c.vmInformer.HasSynced)
+	cache.WaitForCacheSync(stopCh, c.poolInformer.HasSynced, c.vmInformer.HasSynced, c.vmiInformer.HasSynced, c.revisionInformer.HasSynced)
```
Comment (acardace) [inline pkg/virt-controller/watch/pool.go:429]: You're waiting for the vmiInformer twice here, just a copy-paste error probably.

File: pkg/virt-controller/watch/pool_test.go [hunk 2]
```diff
@@ -103,6 +106,7 @@ var _ = Describe("Pool", func() {
 			go vmiInformer.Run(stop)
 			go vmInformer.Run(stop)
 			go poolInformer.Run(stop)
+			go crInformer.Run(stop)
 			Expect(cache.WaitForCacheSync(stop, vmiInformer.HasSynced, vmInformer.HasSynced, poolInformer.HasSynced)).To(BeTrue())
```
Comment (acardace) [inline pkg/virt-controller/watch/pool_test.go]: Shouldn't you also wait for the `crInformer` here?

**Comment by acardace** [comment]
I guess this comment https://github.com/kubevirt/kubevirt/pull/7139/commits/8a3de5036adbf3a70f189a3743e70fe301597a90#diff-23ab0c5d26111ec6a9c2ab2e922be85716f9ab1e45014f6a93d8958a5ad46b13R161 is outdated now, it should probably be removed?

**Review by acardace — APPROVED** [review]

File: pkg/virt-controller/watch/pool.go [hunk 3]
```diff
@@ -712,18 +703,25 @@ func (c *PoolController) scaleOut(pool *poolv1.VirtualMachinePool, count int) er
 
 	newNames := calculateNewVMNames(count, pool.Name, pool.Namespace, c.vmInformer.GetStore())
 
+	if len(newNames) == 0 {
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply pkg/virt-controller/watch/pool.go:706]: yep, it is impossible. i'm removing this

File: pkg/virt-controller/watch/pool.go [hunk 4]
```diff
@@ -769,32 +767,41 @@ func (c *PoolController) scaleOut(pool *poolv1.VirtualMachinePool, count int) er
...
+			return &syncErrorImpl{fmt.Errorf("Error during scale in: %v", err), FailedScaleInReason}, false
 		}
 	}
 
-	return nil
+	return nil, false
 }
 
-func (c *PoolController) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutdatedList []*virtv1.VirtualMachine, vmHash string, vmiHash string) error {
+func (c *PoolController) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutdatedList []*virtv1.VirtualMachine) error {
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply pkg/virt-controller/watch/pool.go:793]: we need the list to know which VMs to act on. The list itself isn't mutated, but it's used

**Comment by davidvossel** [comment]
updated to address feedback and fix merge conflict

File: pkg/virt-controller/watch/pool_test.go (hunk 2 above)
Comment (davidvossel) [reply pkg/virt-controller/watch/pool_test.go]: yes! fixed

File: pkg/virt-controller/watch/pool_test.go [hunk 5]
```diff
@@ -275,14 +348,18 @@ var _ = Describe("Pool", func() {
...
 			addVM(vm)
 			addVMI(vmi, true)
+			addCR(oldPoolRevision)
+			addCR(newPoolRevision)
+
+			expectControllerRevisionCreation(newPoolRevision)
 
 			vmiInterface.EXPECT().Delete(gomock.Any(), gomock.Any()).Times(1).Return(nil)
-			vmInterface.EXPECT().Update(gomock.Any()).Times(0).Do(func(arg interface{}) {}).Return(vm, nil)
+			//vmInterface.EXPECT().Update(gomock.Any()).Times(0).Do(func(arg interface{}) {}).Return(vm, nil)
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply pkg/virt-controller/watch/pool_test.go]: fixed, thanks

File: pkg/virt-controller/watch/pool.go (hunk 1 above)
Comment (acardace) [reply pkg/virt-controller/watch/pool.go:427]: oops, guess I missed an 'i'.

File: pkg/virt-controller/watch/pool.go [hunk 6]
```diff
@@ -605,30 +621,102 @@ func injectHashIntoVM(vm *virtv1.VirtualMachine, vmHash string, vmiHash string)
 		vm.Spec.Template.ObjectMeta.Labels = map[string]string{}
 	}
 
-	vm.Labels[virtv1.VirtualMachineTemplateHash] = vmHash
-	vm.Spec.Template.ObjectMeta.Labels[virtv1.VirtualMachineInstanceTemplateHash] = vmiHash
+	vm.Labels[virtv1.VirtualMachinePoolRevisionName] = revisionName
+	vm.Spec.Template.ObjectMeta.Labels[virtv1.VirtualMachinePoolRevisionName] = revisionName
```
Comment (rmohr) [inline pkg/virt-controller/watch/pool.go:625]: Should this be an annotation on the VMI? No concerns on the VM. I just wonder about `expose` commands which would automatically consider all labels (we may filter out kubevirt.io prefixed labels though, let's double check).

File: pkg/virt-controller/watch/pool.go [hunk 7]
```diff
@@ -605,30 +621,102 @@ func injectHashIntoVM(vm *virtv1.VirtualMachine, vmHash string, vmiHash string)
...
+		return "", err
+	}
+	revisionName := getRevisionName(pool)
+
+	_, alreadyExists, err := c.getControllerRevision(pool.Namespace, revisionName)
+	if err != nil {
+		return "", err
+	} else if alreadyExists {
+		// already created
+		return revisionName, nil
```
Comment (rmohr) [inline pkg/virt-controller/watch/pool.go:646]: Maybe call this function `ensureControllerRevision` instead of `createControllerRevision`.

**Review by rmohr — COMMENTED** [review]
Two minor comments.

File: pkg/virt-controller/watch/pool.go (hunk 6 above)
Comment (davidvossel) [reply pkg/virt-controller/watch/pool.go:625]: fixed

File: pkg/virt-controller/watch/pool.go (hunk 7 above)
Comment (davidvossel) [reply pkg/virt-controller/watch/pool.go:646]: fixed

**Pull request merged**